lambda function to re-index items to ES

abaevbog opened the pull request:
Async lambda function invoked directly from dataserver that re-adds items deleted from ES to the dlq to be indexed again.

dstillman commented on template.yaml.j2:
```diff
@@ -67,7 +67,34 @@ Resources:
+      Runtime: nodejs18.x
+      CodeUri: ./
+      MemorySize: 128
+      Timeout: 300
```
This might as well be set to the maximum (900 seconds).

dstillman commented on template.yaml.j2:
```diff
@@ -67,7 +67,34 @@ Resources:
-  
+
+  # Async re-indexer of libraries deleted from ES
+  FullTextReIndexerFunction:
```
Lowercase `I` here and below

dstillman commented on index.mjs:
```diff
@@ -200,3 +205,56 @@ export const dlq = async function (event, context) {
 	}
 };
+
+export const reindexDeletedLibrary = async (event) => {
```
Can remove "Deleted". Theoretically it should only be necessary for deleted libraries, but as far as the function is concerned it's just reindexing a library with what's in S3.

dstillman commented on index.mjs:
```diff
@@ -200,3 +205,56 @@ export const dlq = async function (event, context) {
+		});
+		sqsEventsArray = sqsEventsArray.concat(sqsEvents);
+		continuationToken = s3Objects.NextContinuationToken;
+	} while (continuationToken);
```
Probably no difference, but I think `.IsTruncated` is the [more canonical](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ListingKeysUsingAPIs.html) way to do this.

dstillman commented on index.mjs:
```diff
@@ -200,3 +205,56 @@ export const dlq = async function (event, context) {
+			};
+			return { Id: entry.Key.replace("/", "-"), MessageBody: JSON.stringify(message) };
+		});
+		sqsEventsArray = sqsEventsArray.concat(sqsEvents);
```
There could be tens or hundreds of thousands of files in a library, so we don't want to combine these all in one array. They should just get sent out in batches as they're available, and then it can get more from S3 as necessary.

dstillman commented on index.mjs:
```diff
@@ -200,3 +205,56 @@ export const dlq = async function (event, context) {
 };
+
+export const reindexDeletedLibrary = async (event) => {
+	const libraryId = event.libraryID;
```
House style: `libraryID`

dstillman commented on index.mjs:
```diff
@@ -200,3 +205,56 @@ export const dlq = async function (event, context) {
+						},
+					}]
+			};
+			return { Id: entry.Key.replace("/", "-"), MessageBody: JSON.stringify(message) };
```
House style: Separate over multiple lines (other places too)

dstillman commented on index.mjs:
```diff
@@ -200,3 +213,121 @@ export const dlq = async function (event, context) {
+	const reindexRequestDate = event.date;
+	var isTruncated = true;
+	let sqsSendEventPromisses = [];
+	let sqs = new SQSClient();
```
Move down to above the code where it's used

dstillman commented on index.mjs:
```diff
@@ -200,3 +213,121 @@ export const dlq = async function (event, context) {
+export const reindexLibrary = async (event) => {
+	const libraryID = event.libraryID;
+	const reindexRequestDate = event.date;
+	var isTruncated = true;
```
Should move down to above the block where this is used, but we can just use a do/while and avoid a separate variable.

dstillman commented on index.mjs:
```diff
@@ -200,3 +213,121 @@ export const dlq = async function (event, context) {
+	const libraryID = event.libraryID;
+	const reindexRequestDate = event.date;
+	var isTruncated = true;
+	let sqsSendEventPromisses = [];
```
`Promisses` → `Promises`, and move down to above the code where it's used

dstillman commented on index.mjs:
```diff
@@ -200,3 +213,121 @@ export const dlq = async function (event, context) {
+				Entries: batch
+			});
+			sqsSendEventPromisses.push(sqs.send(command));
+		}
```
Why two separate loops? Can just get 10 events at a time and send them off.

dstillman commented on template.yaml.j2:
```diff
@@ -67,7 +67,34 @@ Resources:
+    Type: 'AWS::Serverless::Function'
+    Properties:
+      FunctionName: {{ reindexer_function_name }}
+      Handler: index.reindexDeletedLibrary
```
This need to be updated.

dstillman commented on index.mjs:
```diff
@@ -208,18 +212,58 @@ export const dlq = async function (event, context) {
 
 export const reindexLibrary = async (event) => {
 	const libraryID = event.libraryID;
+	const reindexRequestDate = event.date;
```
What is `event.date`? I'm not seeing that in https://github.com/zotero/dataserver/pull/151.

If this is a timestamp, it'd be clearer to call it `reindexRequestTime`. But if the goal is just to uniquely identify a run, shouldn't it just be a UUID or random string?

dstillman commented on index.mjs:
```diff
@@ -208,18 +212,58 @@ export const dlq = async function (event, context) {
+
+	// If we did not exit, for this requestDate the run is not complete.
+	reindexStatus.requestDate = reindexRequestDate;
+	reindexStatus.status = "inProgress";
```
If complete and request identifier isn't the same, we should clear `.items`, right? Otherwise deleted items would never be removed.

dstillman commented on index.mjs:
```diff
@@ -258,8 +302,28 @@ export const reindexLibrary = async (event) => {
+		for (let entry of Contents) {
+			const key = entry.Key;
+			reindexStatus.items[key] = reindexRequestDate;
+		}
```
Is there any reason to store the request identifier for every item? Don't we only care 1) whether a given requested reindex failed to finish and 2) if it did, which items were already processed? So can't we just store the request identifier and the items processed so far?

And on completion, should we not delete the reindex file? What's the value of `Complete`? I'm not totally clear on the circumstances under which an SQS message [can be delivered multiple times](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/standard-queues.html#standard-queues-at-least-once-delivery), but it seems unlikely that the same message would be delivered once the Lambda competes successfully.

dstillman reviewed (commented):
I just realized I had pending comments here. Not sure if they made sense, but let's load this and https://github.com/zotero/dataserver/pull/151 back into RAM and get them merged.

abaevbog commented:
I would like to go over the general logic here to make sure it's all on the right track. 

Our goal is to have this be a long running operation that re-indexes all items from the library to ES, that is initiated asynchronously from dataserver, and we also need a way to also tell when it started and when it was done.

The library table has a new boolean column `deindexed_from_es` (may need a better name) that should be set to true when the library is removed from ES. 
`GET` to `fulltext/reindex` will return indexing status of `indexed`/`indexing`/`deindexed`. `deindexed` if `deindexed_from_es` is true, `indexed` if the count of indexable attachments = count of records in ES, and `indexing` otherwise. 
`POST`  to `fulltext/reindex` will trigger this lambda to begin reindexing and set `deindexed_from_es=false`. Only goes through if initially `deindexed_from_es` is `false`.

Dataserver will trigger reindexing by adding an event to a new queue to make sure that the event is not lost if there is an outage/many retries are needed. This triggers `reindexLibrary` that lists S3 objects in batches and adds each batch to DLQ. If there are too many items and the lambda is about to timeout, the event goes back to the queue to be processed by the next lambda.  There is a temp `_reindex_status` file added to S3 to keep track of which key was seen last, so that the next lambda picks up where the other one left off.

abaevbog commented:
The flip side of this essentially sequential processing is that it might take much longer for a large library to be indexed, so I'm not sure if knowing the status when the entire batch is finished is worth the tradeoff. 

Question: would checking if `Zotero_FullText::countInLibrary($this->objectLibraryID)` is equal to `Zotero_Libraries::countAttachments($this->objectLibraryID)` be a reasonable alternative way to establish that the library has been fully re-indexed? If so, this last change wouldn't be needed and GET to `fulltext/reindex` could just check that

abaevbog committed:
function to reindex library to ES

- added a new FullTextReIndexerQueue SQS to store requests to reindex
the library. Reindexing starts when dataserver adds an event to that queue.
- FullTextReindexerFunction lambda lists all keys in S3 and adds them
to indexing DLQ. It is done in batches and the progress is saved to
_reindexing_status file, so that if another lambda has to finish processing,
it can pick up where the last one stopped.

abaevbog commented:
I revised this to simplify the logic quite a bit and to work in tandem with [the dataserver PR](https://github.com/zotero/dataserver/pull/151). I did some testing with this lambda deployed with lower timeout values, and it worked as expected with my local instances of dataserver and ES

dstillman committed:
Use a constant for reindex S3 batch size instead of config

The page size for the reindex ListObjectsV2 loop doesn't need to vary
per environment, so hardcode it as REINDEX_S3_BATCH_SIZE rather than
exposing it via config.

dstillman committed:
Add missing reindexer_queue_name to sample config

template.yaml.j2 references {{ reindexer_queue_name }} for the reindex
queue, but it was never added to config.env-sample.

dstillman committed:
Capitalize S3 and SQS in reindexer comments

dstillman committed:
Give reindex queue visibility timeout a margin over the Lambda timeout

VisibilityTimeout was equal to the function Timeout (900), leaving no
buffer; raise it to 960 so a message can't become visible to a second
invocation while the first is still running.

dstillman committed:
Rename FullTextReIndexerQueue to FullTextReindexerQueue

Match the casing of FullTextReindexerFunction and reindexer_function_name.
